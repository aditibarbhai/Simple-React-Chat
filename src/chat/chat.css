.error {
    color: #f44336;
}

.selected {
    background-color: #00000014;
}

.full-width {
    width: 100%;
}

.create-username-wrapper {
    margin: auto;
    text-align: center;
    width: 50vw;
}

#createUsername .form-control {
    display: block;
    margin: 20px auto;
    width: 50%;
}

.users {
    background-color: #eeeeee;
    float: left;
    height: 99vh;
    width: 20vw;
    padding-left: 20px;
    box-shadow: 0 4px 8px 0 rgba(0, 0, 0, 0.2), 0 6px 20px 0 rgba(0, 0, 0, 0.19);
}

.users .users-list {
    background-color: #ffffff;
}

.conversation {
    float: right;
    height: 99vh;
    width: 80vw;
}