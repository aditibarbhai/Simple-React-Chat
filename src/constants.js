module.exports = {
    PORT: 8080,
    connection: 'connection',
    disconnect: 'disconnect',
    CREATE_USER: 'CREATE_USER',
    SEND_MESSAGE: 'SEND_MESSAGE',
    RECEIVE_MESSAGE: 'RECEIVE_MESSAGE',
    WHISPER: 'WHISPER',
    USERNAMES: 'USERNAMES'
};