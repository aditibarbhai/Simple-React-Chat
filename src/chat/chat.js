import React, { Component } from 'react';
import { connect } from 'react-redux';
import io from 'socket.io-client';
import constants from '../constants';
import PersonIcon from '@material-ui/icons/Person';
import PersonOutlineIcon from '@material-ui/icons/PersonOutline';
import AccountCircle from '@material-ui/icons/AccountCircle';
import { 
    CssBaseline,
    List, 
    ListItem, 
    ListItemIcon, 
    ListItemText,
    FormControl,
    FormHelperText,
    Input,
    InputAdornment,
    Button
} from '@material-ui/core';

import { 
    updateSelectedUser, 
    createUsername, 
    updateUsers, 
    updateError,
    updateSocket,
    updateMessages
} from '../actions/chat-actions';
import Conversation from '../conversation/conversation';
import './chat.css';

export class Chat extends Component {
    constructor(props) {
        super(props);

        this.state = {
            username: ''
        };

        this.handleUsernameChange = this.handleUsernameChange.bind(this);
        this.handleCreateUsernameClick = this.handleCreateUsernameClick.bind(this);
        this.selectUser = this.selectUser.bind(this);
        this.updateUsersCallback = this.updateUsersCallback.bind(this);
        this.receiveMessageCallback = this.receiveMessageCallback.bind(this);
        this.createUserCallback = this.createUserCallback.bind(this);
        
        const socket = io(`http://localhost:${constants.PORT}`);
        
        socket.on(constants.USERNAMES, this.updateUsersCallback);
        socket.on(constants.RECEIVE_MESSAGE, this.receiveMessageCallback);
        
        props.updateSocket(socket);
    }

    updateUsersCallback(usernames) {
        this.props.updateUsers(usernames);
    }

    receiveMessageCallback(message) {
        this.props.updateMessages(message);
    }

    selectUser(event) {
        this.props.updateSelectedUser(event.target.textContent);
    }

    renderUserList(users) {
        let connectedUsers = users.filter(user => user !== this.props.myUsername);

        if (connectedUsers.length > 0) {
            return connectedUsers.map((user, index) => {
                let selected = false,
                    icon = <PersonOutlineIcon />

                if (user === this.props.selectedUser) {
                    selected = true;
                    icon = <PersonIcon />;
                }

                const styles = selected ? 'user selected' : 'user';

                return (
                    <ListItem selected={selected} className={styles} key={user + index} button>
                        <ListItemIcon>
                            { icon }
                        </ListItemIcon>
                        <ListItemText onClick={this.selectUser} primary={user} />
                    </ListItem>
                );
            });
        } else {
            return (
                <ListItem className="user" button>
                    <ListItemText primary="No other users connected" />
                </ListItem>
            );
        }
    }

    handleUsernameChange(event) {
        this.setState({
            username: event.target.value.trim()
        });
    }

    createUserCallback(response) {
        if (response.created) {
            this.props.createUsername(this.state.username);
        } else {
            this.props.updateError({
                visible: true,
                message: response.error
            });
        }
    }

    handleCreateUsernameClick(event) {
        event.preventDefault();
        if (this.state.username) {
            this.props.socket.emit(constants.CREATE_USER, this.state.username, this.createUserCallback);
        } else {
            this.props.updateError({
                visible: true,
                message: 'Username is required!'
            });
        }
    }

    render() {
        return (
            <div>
                <CssBaseline />
                {
                    !this.props.myUsername &&
                    <div className="create-username-wrapper">
                        <h1> Welcome to SimpleChat! </h1>
                        <p> To get started, create a username: </p>
                        <form id="createUsername">
                            <FormControl className="form-control" >
                                <Input
                                    id="username"
                                    className="full-width" 
                                    startAdornment={
                                        <InputAdornment position="start">
                                            <AccountCircle />
                                        </InputAdornment>
                                    }
                                    value={this.state.username}
                                    onChange={this.handleUsernameChange}
                                    error={this.props.error.visible}
                                />
                                {
                                    this.props.error.visible &&
                                    <FormHelperText error={this.props.error.visible} > 
                                        { this.props.error.message } 
                                    </FormHelperText>
                                }
                            </FormControl>
                            <FormControl className="form-control">
                                 <Button 
                                    className="full-width" 
                                    onClick={this.handleCreateUsernameClick} 
                                    type="submit"
                                    variant="outlined" 
                                >
                                    Create
                                </Button>
                            </FormControl>
                        </form>
                    </div>
                }
                {
                    this.props.myUsername &&
                    <div className="chat-wrapper">
                        <div className="users">
                            <h2>Connected Users </h2>
                            <List className="users-list" component="nav">
                                { this.renderUserList(this.props.users) }
                            </List>
                        </div>
                        <div className="conversation">
                            {
                                this.props.selectedUser &&
                                <Conversation />
                            }
                        </div>
                    </div>
                }
            </div>
        );
    }
}

Chat.propTypes = {};

const mapStateToProps = state => ({
    selectedUser: state.selectedUser,
    myUsername: state.myUsername,
    users: state.users,
    error: state.error,
    socket: state.socket
});

const mapDispatchToProps = dispatch => ({
    updateSelectedUser: user => dispatch(updateSelectedUser(user)),
    createUsername: username => dispatch(createUsername(username)),
    updateUsers: users => dispatch(updateUsers(users)),
    updateError: error => dispatch(updateError(error)),
    updateSocket: socket => dispatch(updateSocket(socket)),
    updateMessages: message => dispatch(updateMessages(message))
});

export default connect(mapStateToProps, mapDispatchToProps)(Chat);