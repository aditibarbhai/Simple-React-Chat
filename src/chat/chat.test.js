import React from 'react';
import renderer from 'react-test-renderer';
import { shallow, mount } from 'enzyme';
import { Chat } from './chat';
import constants from '../constants';

jest.mock('../conversation/conversation', () => 'Conversation');

describe('Chat', () => {
    let props = {}, wrapper;

    beforeEach(() => {
        props = {
            selectedUser: 'testUser',
            myUsername: 'myTestUsername',
            users: ['testUser', 'myTestUsername'],
            error: {
                visible: false,
                message: ''
            },
            socket: {
                emit: jest.fn(),
                on: jest.fn()
            },
            updateSelectedUser: jest.fn(),
            updateSocket: jest.fn(),
            updateError: jest.fn(),
            createUsername: jest.fn(),
            updateUsers: jest.fn(),
            updateMessages: jest.fn()
        };

        wrapper = shallow(<Chat {...props} />);
    });

    afterEach(() => {
        jest.clearAllMocks();
    });

    it ('should render without errors', () => {
        const tree = renderer
            .create(<Chat {...props} />)
            .toJSON();

        expect(tree).toMatchSnapshot();
    });

    it ('should render create user form if no username', () => {
        props.myUsername = '';
        props.error.visible = true;
        const tree = renderer
            .create(<Chat {...props} />)
            .toJSON();

        expect(tree).toMatchSnapshot();
    });

    it ('selectUser should update selected user', () => {
        wrapper.instance().selectUser({ target: { textContent: 'testUser1' } });

        expect(props.updateSelectedUser).toHaveBeenCalledWith('testUser1');
    });

    it ('handleUsernameChange updates state.username', () => {
        expect(wrapper.state().username).toBe('');

        wrapper.instance().handleUsernameChange({ target: { value: 'testUsername' } });

        expect(wrapper.state().username).toBe('testUsername');
    });

    it ('handleCreateUsernameClick should emit CREATE_USER action', () => {
        wrapper.setState({ username: 'myTestUsername' });
        wrapper.instance().handleCreateUsernameClick({ preventDefault: jest.fn() });

        expect(props.socket.emit).toHaveBeenCalledWith(constants.CREATE_USER, 'myTestUsername', expect.any(Function));
    });

    it ('handleCreateUsernameClick call updateError if no username entered', () => {
        wrapper.instance().handleCreateUsernameClick({ preventDefault: jest.fn() });

        expect(props.updateError).toHaveBeenCalledWith({
            visible: true,
            message: 'Username is required!'
        });
    });

    it ('createUserCallback should call createUsername if successful', () => {
        wrapper.setState({ username: 'testUsername' });
        wrapper.instance().createUserCallback({ created: true });

        expect(props.createUsername).toHaveBeenCalledWith('testUsername');
    });

    it ('createUserCallback should call updateError if error', () => {
        wrapper.instance().createUserCallback({ error: 'test error' });

        expect(props.updateError).toHaveBeenCalledWith({
            visible: true,
            message: 'test error'
        });
    });

    it ('updateUsersCallback calls updateUsers', () => {
        wrapper.instance().updateUsersCallback([]);

        expect(props.updateUsers).toHaveBeenCalledWith([]);
    });

    it ('receiveMessageCallback calls updateUsers', () => {
        wrapper.instance().receiveMessageCallback({
            message: 'test',
            to: 'myUsername',
            from: 'testUser'
        });
        
        expect(props.updateMessages).toHaveBeenCalledWith({
            message: 'test',
            to: 'myUsername',
            from: 'testUser'
        });
    });
});