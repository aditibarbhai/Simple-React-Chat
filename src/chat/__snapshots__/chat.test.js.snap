// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Chat should render create user form if no username 1`] = `
<div>
  <div
    className="create-username-wrapper"
  >
    <h1>
       Welcome to SimpleChat! 
    </h1>
    <p>
       To get started, create a username: 
    </p>
    <form
      id="createUsername"
    >
      <div
        className="MuiFormControl-root-62 form-control"
      >
        <div
          className="MuiInput-root-66 MuiInput-error-71 MuiInput-formControl-67 MuiInput-underline-70 full-width"
        >
          <div
            className="MuiInputAdornment-root-79 MuiInputAdornment-positionStart-80"
          >
            <svg
              aria-hidden="true"
              className="MuiSvgIcon-root-19"
              color={undefined}
              focusable="false"
              viewBox="0 0 24 24"
            >
              <g>
                <path
                  d="M12 2C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2zm0 3c1.66 0 3 1.34 3 3s-1.34 3-3 3-3-1.34-3-3 1.34-3 3-3zm0 14.2c-2.5 0-4.71-1.28-6-3.22.03-1.99 4-3.08 6-3.08 1.99 0 5.97 1.09 6 3.08-1.29 1.94-3.5 3.22-6 3.22z"
                />
              </g>
            </svg>
          </div>
          <input
            aria-invalid={true}
            aria-required={false}
            autoComplete={undefined}
            autoFocus={undefined}
            className="MuiInput-input-74"
            defaultValue={undefined}
            disabled={false}
            id="username"
            name={undefined}
            onBlur={[Function]}
            onChange={[Function]}
            onFocus={[Function]}
            onKeyDown={undefined}
            onKeyUp={undefined}
            placeholder={undefined}
            readOnly={undefined}
            required={undefined}
            rows={undefined}
            type="text"
            value=""
          />
        </div>
        <p
          className="MuiFormHelperText-root-82 MuiFormHelperText-error-83"
        >
          
        </p>
      </div>
      <div
        className="MuiFormControl-root-62 form-control"
      >
        <button
          className="MuiButtonBase-root-15 MuiButton-root-86 MuiButton-flat-90 MuiButton-outlined-93 full-width"
          disabled={false}
          onBlur={[Function]}
          onClick={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          onKeyUp={[Function]}
          onMouseDown={[Function]}
          onMouseLeave={[Function]}
          onMouseUp={[Function]}
          onTouchEnd={[Function]}
          onTouchMove={[Function]}
          onTouchStart={[Function]}
          tabIndex="0"
          type="submit"
        >
          <span
            className="MuiButton-label-87"
          >
            Create
          </span>
          <span
            className="MuiTouchRipple-root-55"
          />
        </button>
      </div>
    </form>
  </div>
  
</div>
`;

exports[`Chat should render without errors 1`] = `
<div>
  <div
    className="chat-wrapper"
  >
    <div
      className="users"
    >
      <h2>
        Connected Users 
      </h2>
      <nav
        className="MuiList-root-1 MuiList-padding-2 users-list"
      >
        <div
          className="MuiButtonBase-root-15 MuiListItem-root-5 MuiListItem-default-8 MuiListItem-gutters-12 MuiListItem-button-13 user selected"
          onBlur={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          onKeyUp={[Function]}
          onMouseDown={[Function]}
          onMouseLeave={[Function]}
          onMouseUp={[Function]}
          onTouchEnd={[Function]}
          onTouchMove={[Function]}
          onTouchStart={[Function]}
          role="button"
          selected={true}
          tabIndex="0"
        >
          <svg
            aria-hidden="true"
            className="MuiSvgIcon-root-19 MuiListItemIcon-root-18"
            color={undefined}
            focusable="false"
            viewBox="0 0 24 24"
          >
            <g>
              <path
                d="M12 12c2.21 0 4-1.79 4-4s-1.79-4-4-4-4 1.79-4 4 1.79 4 4 4zm0 2c-2.67 0-8 1.34-8 4v2h16v-2c0-2.66-5.33-4-8-4z"
              />
            </g>
          </svg>
          <div
            className="MuiListItemText-root-25"
            onClick={[Function]}
          >
            <span
              className="MuiTypography-root-31 MuiTypography-subheading-38 MuiListItemText-primary-28"
            >
              testUser
            </span>
          </div>
          <span
            className="MuiTouchRipple-root-55"
          />
        </div>
      </nav>
    </div>
    <div
      className="conversation"
    >
      <Conversation />
    </div>
  </div>
</div>
`;
